spec/client_spec.rb

```diff
@@ -0,0 +1,89 @@
+RSpec.describe CGIParty::Client do
+  def collect_response_with_status(status)
+    CGIParty::CollectResponse.new(
+      {
+        collect_response: {
+          transaction_id: "transaction_id",
+          progress_status: status
+        }
+      })
+  end
+
+  describe "#poll_collect" do
+    include Savon::SpecHelper
+
+    before { allow_any_instance_of(CGIParty::Client).to receive(:sleep) }
+    let(:order_ref) { "c87ebe55-0d1d-4d37-a5ed-c0b65f88c4e4" }
+    let(:client) { CGIParty::Client.new }
+    before(:all) { savon.mock! }
+    after(:all) { savon.unmock! }
+
+
+    it "must call authenticate and collect" do
+      savon.expects(:collect).with(message: :any).returns(File.read("spec/fixtures/collect_response.xml"))
+
+      client.poll_collect(order_ref) {}
+    end
+
+    it "must timeout" do
+      expect(client).to receive(:collect).ordered.once # Initial collect
+      expect(client).to receive(:polling_duration).and_return(CGIParty.config.collect_polling_timeout)
+      expect(client).to receive(:collect).ordered.never
+
+      client.poll_collect(order_ref) {}
+    end
+
+    it "must call collect until authentication is considered finished" do
+      expect(client).to receive(:collect).exactly(3).times
```

Comment:
I'd much rather use `allow(client).to receive(:collect).and_return()` and then use `expect(client).to have_received(:collect).exactly(3).times`. It reads more naturally.